# Preamble

This notebook is meant to be an example of how to use Jupyter notebooks to document and illustrate applied research in machine learning (which is these days referred to as _data science_). 

I am following to a great extent the great tutorials/walkthroughs on the book (which I *highly* recommend you get a copy of)

```
Hands-On Machine Learning with Scikit-Learn and Tensorflow
Aurelien Geron
O'Reilly Media, 2017
```

the notebooks for the book [can be found here](https://github.com/ageron/handson-ml).

In [1]:
import numpy as np
import os

# To plot figures
%matplotlib inline
import matplotlib as mpl
import matplotlib.pyplot as plt
mpl.rc('axes', labelsize=14)
mpl.rc('xtick', labelsize=12)
mpl.rc('ytick', labelsize=12)

In [2]:
PROJECT_ROOT_DIR="../.."

# Get the Data

In [3]:
import pandas as pd

We load the training dataset directly as a `DataFrame`, note that we're doing so from a compressed file (using `gzip`)

In [4]:
X_data_path = os.path.join(PROJECT_ROOT_DIR, 'X_preprocessed_data.csv.gz')

In [5]:
# See Issue #6 so we do not have to transpose the dataset
df_X = pd.read_csv(X_data_path, header=None)
df_X.head()

0         1        2       3          4    5    6    7    8    9   ...   \
0   15.07   15.1200   14.630   14.75  8407500.0  0.0  1.0  0.0  0.0  0.0 ...    
1   81.22   81.9300   80.940   81.89   296853.0  0.0  0.0  0.0  0.0  0.0 ...    
2   78.24   79.0700   78.125   79.07  4632684.0  0.0  0.0  0.0  0.0  0.0 ...    
3  236.64  238.6924  235.750  238.16   552207.0  0.0  0.0  0.0  0.0  0.0 ...    
4   89.04   89.4800   88.910   89.16   554948.0  0.0  0.0  0.0  0.0  0.0 ...    

   500  501  502  503  504  505  506  507  508  509  
0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
1  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
2  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
3  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
4  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  

[5 rows x 510 columns]

In [6]:
df_X.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 735000 entries, 0 to 734999
Columns: 510 entries, 0 to 509
dtypes: float64(510)
memory usage: 2.8 GB


The `describe()` method allows us to obtain some insight on the structure of the dataset we are working with

In [7]:
df_X.describe()

0              1              2              3    \
count  735000.000000  735000.000000  735000.000000  735000.000000   
mean       58.352279      58.808949      57.971927      58.439323   
std        53.207832      53.659590      52.923152      53.365529   
min         2.580000       2.600000       2.580000       2.590000   
25%        32.307500      32.687500      32.095000      32.260000   
50%        46.020000      46.365000      45.660000      46.050000   
75%        68.280000      68.923550      67.549275      68.152500   
max       703.000000     708.000000     698.900000     705.620000   

                4              5              6              7    \
count  7.350000e+05  735000.000000  735000.000000  735000.000000   
mean   4.813515e+06       0.002381       0.001905       0.002381   
std    1.221788e+07       0.048737       0.043602       0.048737   
min    8.842500e+04       0.000000       0.000000       0.000000   
25%    1.021299e+06       0.000000       0.000000       0.000000   
50%    2.059844e+06       0.000000       0.000000       0.000000   
75%    4.449050e+06       0.000000       0.000000       0.000000   
max    2.317716e+08       1.000000       1.000000       1.000000   

                 8              9        ...                  500  \
count  735000.000000  735000.000000      ...        735000.000000   
mean        0.001905       0.001905      ...             0.001905   
std         0.043602       0.043602      ...             0.043602   
min         0.000000       0.000000      ...             0.000000   
25%         0.000000       0.000000      ...             0.000000   
50%         0.000000       0.000000      ...             0.000000   
75%         0.000000       0.000000      ...             0.000000   
max         1.000000       1.000000      ...             1.000000   

                 501            502            503            504  \
count  735000.000000  735000.000000  735000.000000  735000.000000   
mean        0.002381       0.002381       0.002381       0.001905   
std         0.048737       0.048737       0.048737       0.043602   
min         0.000000       0.000000       0.000000       0.000000   
25%         0.000000       0.000000       0.000000       0.000000   
50%         0.000000       0.000000       0.000000       0.000000   
75%         0.000000       0.000000       0.000000       0.000000   
max         1.000000       1.000000       1.000000       1.000000   

                 505            506            507            508  \
count  735000.000000  735000.000000  735000.000000  735000.000000   
mean        0.002381       0.002381       0.001905       0.002381   
std         0.048737       0.048737       0.043602       0.048737   
min         0.000000       0.000000       0.000000       0.000000   
25%         0.000000       0.000000       0.000000       0.000000   
50%         0.000000       0.000000       0.000000       0.000000   
75%         0.000000       0.000000       0.000000       0.000000   
max         1.000000       1.000000       1.000000       1.000000   

                 509  
count  735000.000000  
mean        0.001905  
std         0.043602  
min         0.000000  
25%         0.000000  
50%         0.000000  
75%         0.000000  
max         1.000000  

[8 rows x 510 columns]

I am not sure what are the 510 columns in the dataset, it looks to me a bit excessive (or it has already been processed into sequences for the RNN?). In any case, it would be good to have a notebook like this one

```
https://github.com/ageron/handson-ml/blob/master/02_end_to_end_machine_learning_project.ipynb
```

so we can study the data before applying any algorithms to it.